runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -591,13 +602,40 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
         ctx.close();
     }
 
+    /**
+     * Sends a ping to the server so that the pong response will keep
+     * the channel active in order to avoid a ReadTimeout exception that will
+     * close the channel.
+     */
+    private void keepAlive() {
+        if (!channel.isOpen()) {
+            log.warn("keepAlive: channel not open, skipping ping. ");
+            return;
+        }
+        channel.writeAndFlush(new CorfuMsg(CorfuMsgType.PING), channel.voidPromise());
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (keepAliveTask != null) {
+            log.debug("channelInactive: Canceling keep-alive task");
+            keepAliveTask.cancel(true);
+            keepAliveTask = null;
+        }
+
+        // Should the handshake handler also do this?
+        this.channelInactive(ctx);
+    }
+
     @Override
     public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
         if (evt.equals(ClientHandshakeEvent.CONNECTED)) {
             // Handshake successful. Complete the connection future to allow
             // clients to proceed.
             channel = ctx.channel();
             connectionFuture.complete(null);
+            keepAliveTask = ctx.executor().scheduleAtFixedRate(this::keepAlive, keepAliveInterval,
```

Comment:
The scheduleAtFixedRate can fail silently. This blocks the thread pool.
You can use the `runSansThrow` method to wrap your method so that the Exceptions are logged but does not crash the thread.
Please refer:
https://github.com/CorfuDB/CorfuDB/blob/c1348b8a88d9f7daa3ad4ba3f63096c6e3fd9aa1/infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java#L310-L314

Reply:
Block the thread pool or fail to be rescheduled ?

Reply:
Fail to be rescheduled.

Reply:
I see. Note that these are wrapped executors, I think if an exception happens during execution it will invoke:

@Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        log.error("Exception during channel handling.", cause);
        ctx.close();
    }

Reply:
Even if the threads have an `uncaughtExceptionHandler` registered, the scheduleAtFixedRate does not honour it. I am not sure if the netty exceptionHandler would be invoked or not.

Reply:
If there's an exception then the channel will be closed, reconnecting will re-add the keep-alive task. 
Swallowing the exception is bad, because we want to close the channel on the exception.